### DISCUSSION TITLE
UI updating using observer pattern (MVC style)

### DISCUSSION THREAD

**Discussion by mmontone** [rabbibotton/clog#239] [Ideas]
As a CLOG app gets more complex, there's a need of updating several places in the user interface when the app model changes.

One way of doing it is to keep a reference to the CLOG element that will need to be updated, and when certain functions are called on your model, redraw the element. But that soon gets difficult to keep track of and implement.

A better way IMO is to use the Observer pattern, like we do in Model-View-Controller architecture. Setup change events on our models and trigger a redraw of pertinent parts of the UI. I'm using this approach in one of my applications. For the events I'm using `event-emitter` library. And for updating the UI, I have a simple but very handy macro:

```lisp
(defmacro updating-ui (clog-obj model event &body body)
  "Render CLOG-OBJ evaluating BODY, and refresh CLOG-OBJ when MODEL triggers EVENT."
  `(progn
     ,@body ;; draw ui for the first time
     (event-emitter:on ',event ,model
                       (lambda (&rest args)
                         (declare (ignore args))
                         (setf (text ,clog-obj) "") ;; delete UI elem contents
                         ;; redraw ui:
                         ,@body))))
```

So this is how it is used. Let's suppose we have a TODO list that emits change events whenever it changes:

```lisp
(defclass todo-list (event-emitter:event-emitter)
  ((items :initarg :items
	  :accessor items)))

(defun add-todo (task todo-list)
  (push task (items todo-list))
  (event-emitter:emit :changed todo-list))
```

Then in our UI code we use our `updating-ui` macro when creating the CLOG view for our changing model:

```lisp
(defun create-todo-list (clog-obj todo-list)
  (let ((ul (clog:create-unordered-list clog-obj)))
    (updating-ui ul todo-list :changed
      (dolist (item (items todo-list))
	(clog:create-list-item ul :content item)))))
```
Now every time our todo list changes (by calling `add-todo`) the body of the `ul` passed to `updating-ui` is emptied and then `updating-ui`'s body is reevaluated, redrawing our list.
And that's it. We don't need to do anything else for keeping the todo-list user interface updated.

Hope you find this useful, and please let me know of your thoughts and if you use other techniques for this problem.

**Comment by sabracrolleton** [comment]
What would be the best way to update a graph every five minutes?

**Reply by mmontone** [reply]
If you used the technique I described, then make the graph an event-emitter, and make it emit a change event each five minutes, using cl-cron library for example. But I'm not sure what exactly is your graph and how it is to be displayed.

**Reply by rabbibotton** [reply]
You can also just use sleep in a loop in a thread.

**Comment by rabbibotton** [comment]
Looks great. I often use clog:make-hash-table* and maphash. What is the advantage of using the emitter library?

**Reply by mmontone** [reply]
It lets me work at an event emit/listen protocol level. If I were to implement an event listener using hash-tables, then I would get the event-emitter library.

But you make a good point. Now I'm thinking that using a system wide hash-table  would allow making any object observable, without the need of sub-classing observable objects from event-emitter class.

Also, events machinery for observers could make use of weak pointers, because the registering of event handlers doesn't let collect the view object if the model is not used anymore. 

So, the event-emitter library does the job, and lets me work at a emit/listen protocol level, but there's room for improvement.
